sdks/go/pkg/beam/core/typex/class.go

```diff
@@ -91,16 +92,25 @@ func ClassOf(t reflect.Type) Class {
 // data must be fully serializable. Functions and channels are examples of invalid
 // types. Aggregate types with no universals are considered concrete here.
 func IsConcrete(t reflect.Type) bool {
+	concrete, _ := isConcrete(t, make(map[uintptr]bool))
+	return concrete
+}
+
+// AssertConcrete returns true iff the given type is a valid "concrete" data type and if not,
+// returns false along with an error indicating why the type is not concrete. Concrete
+// data must be fully serializable. Functions and channels are examples of invalid
+// types. Aggregate types with no universals are considered concrete here.
+func AssertConcrete(t reflect.Type) (bool, error) {
 	return isConcrete(t, make(map[uintptr]bool))
 }
 
-func isConcrete(t reflect.Type, visited map[uintptr]bool) bool {
+func isConcrete(t reflect.Type, visited map[uintptr]bool) (bool, error) {
```

Comment:
Consider instead only returning the error. In all non-nil errors, we are returning false, so we might as well avoid the extra bool, then we check nilness in the IsConcrete variant to maintain the boolean there.

Reply:
That sounds good to me. I considered doing the same in `CheckConcrete` and just returning an error, but decided to leave it because otherwise we're technically leaking some (small) responsibility of answering if a type is concrete to the caller which felt wrong.